### ISSUE TITLE
Add possibility to close connection / stream

### ISSUE THREAD

**Issue by kelunik** [amphp/aerys#194]
There should be a possibility to close a connection / stream (for H2) to abort requests if things go wrong. One such example is a reverse proxy, where the remote hangs up before the full response is sent.

**Real World Example**
```php
$response->setHeader("content-type", $proxyResponse->getHeader("content-type") ?? "image/jpeg");
$response->setHeader("cache-control", "public, max-age=31536000");

try {
    yield ByteStream\pipe($proxyResponse->getBody(), $response);
} catch (HttpException $e) {
    // what should I do now?
}
```

**Comment by bwoebi** [comment]
So, you essentially want to introduce a semantic difference between RST and normal Response->end()?

Will that cause any effective difference in browser behavior? Is there a practical benefit to it? (apart from semantic correctness)

**Comment by kelunik** [comment]
I guess a browser detects this as truncated response and won't use or cache it.

**Comment by bwoebi** [comment]
Does it / can you try that out?

I can't find any information how they handle it.

Also, probably useful in HTTP/1.1 to force the close of a connection (to signal an error) instead of terminating the chunked transfer with a `0\r\n\r\n`.

As far as I see, it will need an extra API method, like `Response->abort() : void`. Or can you think of a solution not requiring an interface change?

**Comment by kelunik** [comment]
Yes, it obviously needs an interface change.

> Also, probably useful in HTTP/1.1 to force the close of a connection (to signal an error) instead of terminating the chunked transfer with a 0\r\n\r\n.

Yes, that's the main intention.

**Comment by kelunik** [comment]
I think with the new interface in `0.8.x` we shouldn't add an explicit `abort` method to request / response. Rather, the response should automatically be aborted if reading from the body stream throws an exception in the response writer.

**Comment by kelunik** [comment]
Implemented via 90c28263511524ea232a3382a758d0629de9771e and 18b9e80bb483eabb4ac2acf868b02059b7480209.
